src/nnet3/nnet-simple-component.h

```diff
@@ -714,6 +714,19 @@ class NaturalGradientAffineComponent: public AffineComponent {
   // in Info().
   double max_change_scale_stats_;
 
+  /// The following function is only called if max_change_per_sample_ > 0, it returns a
+  /// scaling factor alpha <= 1.0 (1.0 in the normal case) that enforces the
+  /// "max-change" constraint.  "in_products" is the inner product with itself
+  /// of each row of the matrix of preconditioned input features; "out_products"
+  /// is the same for the output derivatives.  gamma_prod is a product of two
+  /// scalars that are output by the preconditioning code (for the input and
+  /// output), which we will need to multiply into the learning rate.
+  /// out_products is a pointer because we modify it in-place.
```

Comment:
As far as my understanding, you are suggesting a uniform max-change scaling code for each updatable component in the trainer code (nnet-training.cc): simply compute l2 norm of parameters in each individual updatable component of an nnet (specifically "delta_nnet_"), comparing them with their own max-change value, and then scale down them individually if necessary. In this case, we don't need the complicated function NaturalGradientAffineComponent::GetScalingFactor(), which is specifically dealing with NaturalGradient things.